iree/compiler/Codegen/Common/ConvertToDestinationPassingStylePass.cpp

```diff
@@ -0,0 +1,345 @@
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Transforms/Passes.h"
+
+namespace mlir {
+namespace iree_compiler {
+
+namespace {
+class ConvertToDestinationPassingStylePass
+    : public ConvertToDestinationPassingStyleBase<
+          ConvertToDestinationPassingStylePass> {
+ public:
+  ConvertToDestinationPassingStylePass() = default;
+  void runOnOperation() override;
+};
+}  // namespace
+
+/// Returns the subview into the buffer that is supposed to be populated with
+/// the `value` of the `flow.dispatch.tensor.store` operation. This can be used
+/// to compute the results in place.
+static Value getTensorLoadOpForTensorStoreOp(OpBuilder &b, Operation *storeOp) {
+  SmallVector<Value, 4> operandsOfLoadOp;
+  auto op = cast<OffsetSizeAndStrideOpInterface>(storeOp);
+  Value target, source;
+  std::tie(source, target) =
+      TypeSwitch<Operation *, std::tuple<Value, Value>>(op)
+          .Case<IREE::Flow::DispatchTensorStoreOp>([&](auto storeOp) {
+            return std::make_tuple(storeOp.value(), storeOp.target());
+          })
+          .Case<tensor::InsertSliceOp>([&](auto storeOp) {
+            return std::make_tuple(storeOp.source(), storeOp.dest());
+          })
+          .Default([](Operation *) {
+            return std::make_tuple<Value, Value>(nullptr, nullptr);
+          });
+  if (!target) return nullptr;
+
+  // Clone the offset, size and stride values. They will be CSE-ed later.
+  Operation *parentOp = storeOp->getParentOp();
+  BlockAndValueMapping indexValMap;
+  llvm::SetVector<Operation *> slice;
+  auto cloneIndexValues = [&](ArrayRef<OpFoldResult> ofrs) {
+    SmallVector<OpFoldResult> clonedVals;
+    for (auto ofr : ofrs) {
+      // Just copy the attributes.
+      if (auto attr = ofr.dyn_cast<Attribute>()) {
+        clonedVals.push_back(attr);
+        continue;
+      }
+      Value val = ofr.get<Value>();
+      // If it is a block argument use the same value.
+      if (val.isa<BlockArgument>()) {
+        clonedVals.push_back(val);
+        continue;
+      }
+      // The slice of ops needed for index computation need to be cloned to
+      // avoid use-def violations. If the value has been cloned already, reuse
+      // that.
+      if (auto lookupVal = indexValMap.lookupOrNull(val)) {
+        clonedVals.push_back(lookupVal);
+        continue;
+      }
+      slice.clear();
+      getBackwardSlice(val, &slice, [&](Operation *sliceOp) {
+        return sliceOp->getParentOp() == parentOp;
+      });
+      for (auto sliceOp : slice) {
+        if (!indexValMap.contains(sliceOp->getResult(0))) {
+          b.clone(*sliceOp, indexValMap);
+        }
+      }
+      if (Operation *definingOp = val.getDefiningOp()) {
+        b.clone(*definingOp, indexValMap);
+      }
+      clonedVals.push_back(indexValMap.lookup(val));
+    }
+    return clonedVals;
+  };
+  SmallVector<OpFoldResult> loadOffsets, loadSizes, loadStrides;
+  loadOffsets = cloneIndexValues(op.getMixedOffsets());
```

Comment:
Why are you cloning these? Is it possible to use `getMixedOffsets()` directly?

Reply:
Some times these offset values might be defined "after" the point at which I need to insert the `flow.dispatch.load.tensor` (or `tensor.extract_slice` in the upstream case). So cloning just makes it easier. Needs to work for dynamic cases as well.